explorer/interpreter/interpreter.cpp

```diff
@@ -1087,7 +1085,23 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
       if (auto rewrite = op.rewritten_form()) {
         return todo_.ReplaceWith(std::make_unique<ExpressionAction>(*rewrite));
       }
-      if (act.pos() != static_cast<int>(op.arguments().size())) {
+      if ((op.op() == Operator::And || op.op() == Operator::Or) &&
+          act.pos() >= 1) {
+        auto operand_value =
+            cast<BoolValue>(*act.results()[act.pos() - 1]).value();
+        if ((op.op() == Operator::Or && operand_value) ||
+            (op.op() == Operator::And && !operand_value) || act.pos() == 2) {
+          // Short-circuit evaluation for 'and'/'or'
+          CARBON_ASSIGN_OR_RETURN(Nonnull<const Value*> value,
+                                  ErrorOr<Nonnull<const Value*>>(
+                                      arena_->New<BoolValue>(operand_value)));
+          return todo_.FinishAction(value);
+        } else {
+          // No short-circuit, evaluate 2nd operand
+          Nonnull<const Expression*> arg = op.arguments()[act.pos()];
+          return todo_.Spawn(std::make_unique<ExpressionAction>(arg));
+        }
+      } else if (act.pos() != static_cast<int>(op.arguments().size())) {
```

Comment:
I don't think we need to duplicate the  logic here:

```suggestion
        }
        // No short-circuit, fall through to evaluate 2nd operand.
      }
      
      if (act.pos() != static_cast<int>(op.arguments().size())) {
```